### DISCUSSION TITLE
How to build Tauri App with SQLite

### DISCUSSION THREAD

**Discussion by zhanglun** [tauri-apps/tauri#5440] [Q&A] [answered]
I use SQLite as database and connect with [diesel](https://diesel.rs/) in my project [Lettura](https://github.com/zhanglun/lettura).

I set the DATABASE_URL environment variable in the .env file 

```
DATABASE_URL=../public/lettura.db
```
and connect database with diesel

```rust
pub fn establish_connection() -> SqliteConnection {
  dotenv().ok();

  let database_url = env::var("DATABASE_URL").expect("DATABASE_URL must be set");
  SqliteConnection::establish(&database_url)
    .expect(&format!("Error connecting to {}", database_url))
}
```

Everything is fine when I run `dev` or `build`. But I can't open the app which bundled.When I check the file `/Volumes/Lettura/Lettura.app/Contents/MacOS/Lettura`, there shows some error.

```
thread 'main' panicked at 'DATABASE_URL must be set: NotPresent', src/db.rs:14:47
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I'm stuck here with this problem.Is that mean the app can't find `DATABASE_URL` ? and I don't know if the SQLite database file has already been created before app opening.

Can someone kindly help me with this?

**Comment by zhanglun** [comment, marked as answer]
hey guy, I have solved this problem. here is my solution,  hope it can help someone in need 😄

The dotenv crate reads the file dynamically at runtime. So it's reasonable that this does not work in production, as there is not dotenv file. So  I declared variable 'DATABASE_URL' in `db.rs` which value is relative to the tauri context's path.

```rs
pub fn establish_connection() -> SqliteConnection {
  let database_url = "./lettura.db";
 
  SqliteConnection::establish(&database_url)
    .expect(&format!("Error connecting to {}", database_url))
}
```

Then do migrations at compile time, [here](https://docs.diesel.rs/1.4.x/diesel_migrations/macro.embed_migrations.html) is the detail

```rs
#[macro_use]
extern crate diesel_migrations;

use diesel_migrations::{embed_migrations};
embed_migrations!("./migrations");

fn main () {
  ...
  let connection = db::establish_connection();

  embedded_migrations::run(&connection).expect("Error migrating");
  ...
}

```

**Reply by zahrulsch** [reply]
Nice, I've already looking for the solution and finally found it here

**Comment by toanmx0101** [comment]
After few hours debug, here is my solution with SQLite, hope it will help.

db.rs

```
pub fn establish_connection() -> SqliteConnection {
    dotenv().ok();

    let database_url = env::var("DATABASE_URL").unwrap_or_else(|_| {
        let mut path = env::current_exe().unwrap();
        path.pop();
        path.push("dbname.sqlite");
        path.to_str().unwrap().to_string()
    });

    // Establish the connection to the SQLite database
    SqliteConnection::establish(&database_url)
        .unwrap_or_else(|_| panic!("Error connecting to {}", database_url))
}

```

main.rs

```
use diesel_migrations::{embed_migrations, EmbeddedMigrations, MigrationHarness};

const MIGRATIONS: EmbeddedMigrations = embed_migrations!();

fn main() {
    let connection = &mut establish_connection();  // Your function to establish an SQLite connection

    // Run the embedded migrations on startup
    connection.run_pending_migrations(MIGRATIONS).expect("Failed to run migrations");
   ...
}
```
